lib/ClangImporter/ImportDecl.cpp

```diff
@@ -112,6 +112,11 @@ static AccessLevel getOverridableAccessLevel(const DeclContext *dc) {
   return (dc->getSelfClassDecl() ? AccessLevel::Open : AccessLevel::Public);
 }
 
+static bool privateOrProtected(const clang::Decl *decl) {
```

Comment:
We already have `importer::shouldSuppressDeclImport`. If you put this check there, you can nuke all of the extra stuff you've added here, and you'll handle the lazy importing cases for free.

Reply:
Thanks for the pointer -- doing this in a single place sounds appealing!

Unfortunately, when I tried this (i.e. returning true for private and protected members in `importer::shouldSuppressDeclImport` and removing the other code I had for this), it didn't work -- my tests started failing because the private members had become visible.

As far as I can tell, this is because `ClangImporter::Implementation::lookupValue` is only called for top-level declarations -- e.g. for the `PublicPrivate` class in my test. The class itself is, of course, imported. When my test then tries to access private members of `PublicPrivate`, I think Swift goes looking for them directly in the AST, which still contains these members. (I should say I'm new to the codebase, so my understanding may be wrong.)

I think this could be resolved by putting a call to `shouldSuppressDeclImport` in `ClangImporter::Implementation::importDeclImpl`. I'm not sure if this would have undesired consequences though -- maybe there's a reason this isn't being done yet? IOW, maybe we want to prevent the declarations from being added to the lookup table, but we still want them to be in the AST -- do you know?

Reply:
Having thought about this again, I've realized that while maybe I can't put `shouldSuppressDeclImport` in `ClangImporter::Implementation::importDeclImpl`, I could definitely do the access specifier check there -- see the commit I just pushed.

WDYT?